vng_api_common/middleware.py

```diff
@@ -187,9 +187,16 @@ def client_id(self) -> Union[str, None]:
                 code='jwt-decode-error'
             )
 
+        return payload
```

Comment:
This is a big security vector, since the JWT is never cryptographically checked - see the `verify=False` kwarg on line 182.

We _must_ validate the JWT, which happens in the original `client_id` property.

I think something along these lines is much more correct:

```python
    @property
    def payload(self):
        if self.encoded is None:
            return None

        if not hasattr(self, '_payload'):
            # decode the JWT and validate it

            # jwt check
            try:
                payload = jwt.decode(self.encoded, verify=False)
            except jwt.DecodeError:
                logger.info("Invalid JWT encountered")
                raise PermissionDenied(
                    _('JWT could not be decoded. Possibly you made a copy-paste mistake.'),
                    code='jwt-decode-error'
                )

            # get client_id
            try:
                client_id = payload['client_id']
            except KeyError:
                try:
                    header = jwt.get_unverified_header(self.encoded)
                except jwt.DecodeError:
                    logger.info("Invalid JWT encountered")
                    raise PermissionDenied(
                        _('JWT could not be decoded. Possibly you made a copy-paste mistake.'),
                        code='jwt-decode-error'
                    )
                else:
                    try:
                        client_id = header['client_identifier']
                    except KeyError:
                        raise PermissionDenied(
                            'Client identifier is niet aanwezig in JWT',
                            code='missing-client-identifier'
                        )
                    else:
                        logger.warning(_('The support of authorization old format will be terminated soon. '
                                         'Please use a new format with the separate Authorization Component'))

            # find client_id in DB and retrieve it's secret
            try:
                jwt_secret = JWTSecret.objects.get(identifier=client_id)
            except JWTSecret.DoesNotExist:
                raise PermissionDenied(
                    'Client identifier bestaat niet',
                    code='invalid-client-identifier'
                )
            else:
                key = jwt_secret.secret

            # check signature of the token
            try:
                payload = jwt.decode(self.encoded, key, algorithms='HS256')
            except jwt.InvalidSignatureError as exc:
                logger.exception("Invalid signature - possible payload tampering?")
                raise PermissionDenied(
                    'Client credentials zijn niet geldig',
                    code='invalid-jwt-signature'
                )

            if 'client_id' not in payload:
                payload['client_id'] = client_id

            self._payload = payload

        return self._payload

    @property
    def client_id(self):
        if not self.payload:
            return None

        return self.payload['client_id']
```

1. We provide a (cached) `payload` atrribute, which may return `None` if there's no data to decode
2. If something is wrong during decoding, a `PermissionDenied` is raised
3. If we have an old style JWT, the `client_id` claim is moved from the header to the (now verified) payload
4. The `client_id` is fetched from the payload (which is cached already)